### PR TITLE
Add a set of tests for LLVM 19

### LINKED ISSUE #107681: `.next().unwrap_unchecked()` on Iterator doesn't optimize as expected
Update: per https://github.com/rust-lang/rust/issues/107681#issuecomment-1911246892, this will be fixed with LLVM 19

(Context: this is a minimized example of the behaviour I hit while working on https://github.com/rust-lang/rust/pull/107634/files#r1096624164)

I tried this code: <https://rust.godbolt.org/z/EjWscs1e9>

```rust
pub unsafe fn demo_std(x: &mut Copied<Iter<'_, u32>>) -> u32 {
    x.next().unwrap_unchecked()
}
```

I expected to see this happen: the code would read and bump the pointer in the iterator without checking the end, something like
```nasm
example::demo_std:
        mov     rax, qword ptr [rdi + 8]
        lea     rcx, [rax + 4]
        mov     qword ptr [rdi + 8], rcx
        mov     eax, dword ptr [rax]
        ret
```

Instead, this happened: it still checks the iterator ending condition
```nasm
example::demo_std:
        mov     rax, qword ptr [rdi + 8]
        cmp     rax, qword ptr [rdi]
        je      .LBB0_1
        lea     rcx, [rax + 4]
        mov     qword ptr [rdi + 8], rcx
        mov     eax, dword ptr [rax]
        ret
.LBB0_1:
        ret
```

Dunno where the problem is here -- could be LLVM, could be how the functions are written, could be how MIR handles things...

---

LLVM after optimizing:
```llvm
define noundef i32 @_ZN7example8demo_std17hd41113e4f46707e0E(ptr noalias nocapture noundef align 8 dereferenceable(16) %x) unnamed_addr #0 personality ptr @rust_eh_personality {
start:
  tail call void @llvm.experimental.noalias.scope.decl(metadata !2)
  %0 = getelementptr inbounds { ptr, ptr }, ptr %x, i64 0, i32 1
  %self1.i.i = load ptr, ptr %0, align 8, !alias.scope !5, !nonnull !8, !noundef !8
  %self2.i.i = load ptr, ptr %x, align 8, !alias.scope !5, !nonnull !8, !noundef !8
  %_10.i.i = icmp eq ptr %self1.i.i, %self2.i.i
  br i1 %_10.i.i, label %"_ZN104_$LT$core..iter..adapters..copied..Copied$LT$I$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$4next17h1de76f250ae29b09E.exit", label %bb3.i.i

bb3.i.i:                                          ; preds = %start
  %1 = getelementptr inbounds i32, ptr %self1.i.i, i64 1
  store ptr %1, ptr %0, align 8, !alias.scope !5
  %v.i.i = load i32, ptr %self1.i.i, align 4, !alias.scope !9, !noalias !2, !noundef !8
  br label %"_ZN104_$LT$core..iter..adapters..copied..Copied$LT$I$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$4next17h1de76f250ae29b09E.exit"

"_ZN104_$LT$core..iter..adapters..copied..Copied$LT$I$GT$$u20$as$u20$core..iter..traits..iterator..Iterator$GT$4next17h1de76f250ae29b09E.exit": ; preds = %start, %bb3.i.i
  %.sroa.3.0.i.i = phi i32 [ %v.i.i, %bb3.i.i ], [ undef, %start ]
  %2 = xor i1 %_10.i.i, true
  tail call void @llvm.assume(i1 %2)
  ret i32 %.sroa.3.0.i.i
}
```
Looks like CSE figured out that it's assuming the same thing as the `br` condition in `start`, but nothing moved it up to `start` even though that block dominates the assume?  If it had, then I think the the branch and compare might have properly disappeared?

### LINKED ISSUE #118306: Missed optimization: `_ => 0` generates worse code than `0 => 0, _ => unreachable!()`
Consider the following functions (https://godbolt.org/z/a8r3Tc7TE):

```rust
pub fn faster(input: u64) -> u64 {
  match input % 4 {
    0 => 0,
    1 | 2 => 1,
    3 => 2,
    _ => unreachable!(),
  }
}

pub fn branchy(input: u64) -> u64 {
  match input % 4 {
    1 | 2 => 1,
    3 => 2,
    _ => 0,
  }
}
```

These functions have identical behavior: they map `input` to `input % 4 - (input % 4 / 2)`. In the former case, LLVM generates a nice lookup table for us, but in the latter, it emits an extra branch. The only difference is that I've used `_ => ...` to avoid needing to write an unreachable-by-optimization branch.

If we look at the generated IR (after `-Cpasses=strip,mem2reg,simplifycfg`):

```ll
define i64 @faster(i64 %0) unnamed_addr #0 {
  %2 = urem i64 %0, 4
  switch i64 %2, label %.unreachabledefault [
    i64 0, label %5
    i64 1, label %3
    i64 2, label %3
    i64 3, label %4
  ]

.unreachabledefault:                              ; preds = %1
  unreachable

3:                                                ; preds = %1, %1
  br label %5

4:                                                ; preds = %1
  br label %5

5:                                                ; preds = %1, %4, %3
  %.0 = phi i64 [ 2, %4 ], [ 1, %3 ], [ 0, %1 ]
  ret i64 %.0
}

define i64 @branchy(i64 %0) unnamed_addr #0 {
  %2 = urem i64 %0, 4
  switch i64 %2, label %5 [
    i64 1, label %3
    i64 2, label %3
    i64 3, label %4
  ]

3:                                                ; preds = %1, %1
  br label %5

4:                                                ; preds = %1
  br label %5

5:                                                ; preds = %1, %4, %3
  %.0 = phi i64 [ 2, %4 ], [ 1, %3 ], [ 0, %1 ]
  ret i64 %.0
}
```

The problem is clear: LLVM does not seem to realize that it can trivially transform `branchy` to `faster` here, by observing that the default in the `switch` is only taken when `%2 == 0`.

I suspect this is more LLVM bug than Rust bug, but it feels fixable by a MIR peephole optimization? Unclear. The `_ => 0` code I wrote is an attractive nuisance that I imagine other people writing, too, so perhaps there is value to seeing if this optimization can be made before LLVM.

This bug is also present in Clang, in case someone wants to file an LLVM bug: https://godbolt.org/z/x7rec97E7. It's unclear to me if this is the sort of optimization Clang would do in the frontend instead of in LLVM; could go either-or here, tbh.

### LINKED ISSUE #126585: Two subsequent labels and unnecessary `jmp` with `opt-level=s`
The following code:

```rust
const fn checked_div_round(a: u64, b: u64) -> Option<u64> {
    match b {
        0 => None,
        1 => Some(a),
        // `a / b` is computable and `(a % b) * 2` can not overflow since `b >= 2`. UPDATE <- this is wrong
        b => Some(a / b + if (a % b) * 2 >= b { 1 } else { 0 }),
    }
}
```

Generates two subsequent labels and an unnecessary `jmp` instruction:

```asm
checked_div_round:
        test    rsi, rsi
        je      .LBB0_1
        cmp     rsi, 1
        je      .LBB0_3
        mov     rax, rdi
        xor     edx, edx
        div     rsi
        add     rdx, rdx
        cmp     rdx, rsi
        mov     rdi, rax
        sbb     rdi, -1
.LBB0_3:
        mov     esi, 1
        jmp     .LBB0_4   ; this jump does not seem necessary since `.LBB0_1` is empty
.LBB0_1:
.LBB0_4:
        mov     rax, rsi
        mov     rdx, rdi
        ret
```

When compiled with `-C opt-level=s` using rustc `1.79.0`.

Godbolt: https://godbolt.org/z/fhTq3ssnK

P.S. For context, I am not too familiar with asm. I did not find any similar results while searching through the rustc and llvm issue trackers for "double | subsequent | unnecessary label" or "unnecessary jmp | jump". Please let me know how I can improve the quality of my issue reporting.

### PR DESCRIPTION
**Pull request by dianqk**
Close #107681. Close #118306. Close #126585.

r? compiler

### CODE REVIEW

File: tests/codegen/issues/issue-107681-unwrap_unchecked.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+use std::slice::Iter;
+
+#[no_mangle]
+pub unsafe fn foo(x: &mut Copied<Iter<'_, u32>>) -> u32 {
+    // CHECK-LABEL: @foo(
+    // CHECK-NOT: br
+    // CHECK-NOT: select
+    // CHECK: [[RET:%.*]] = load i32, ptr
+    // CHECK-NEXT: ret i32 [[RET]]
+    x.next().unwrap_unchecked()
```
Comment (DianQK) [inline tests/codegen/issues/issue-107681-unwrap_unchecked.rs:19]: https://godbolt.org/z/86WcqPssP

File: tests/codegen/issues/issue-118306.rs [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+#[no_mangle]
+pub fn branchy(input: u64) -> u64 {
+    // CHECK-LABEL: @branchy(
+    // CHECK-NEXT:  start:
+    // CHECK-NEXT:    [[_2:%.*]] = and i64 [[INPUT:%.*]], 3
+    // CHECK-NEXT:    [[SWITCH_GEP:%.*]] = getelementptr inbounds [4 x i64], ptr @switch.table.branchy, i64 0, i64 [[_2]]
+    // CHECK-NEXT:    [[SWITCH_LOAD:%.*]] = load i64, ptr [[SWITCH_GEP]]
+    // CHECK-NEXT:    ret i64 [[SWITCH_LOAD]]
+    match input % 4 {
```
Comment (DianQK) [inline tests/codegen/issues/issue-118306.rs:17]: https://godbolt.org/z/4zqePjKbo

File: tests/assembly/issues/issue-126585.rs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    // CHECK-NEXT: .cfi_startproc
+    // CHECK-NEXT: cmp
+    // CHECK-NEXT: jbe [[L1:..*]]
+    // CHECK-NOT: je
+    // CHECK-NOT: jmp
+    // CHECK-NOT: .{{.*}}:
+    // CHECK: [[L1]]:
+    // CHECK-NOT: .{{.*}}:
+    // CHECK: ret
+    match b {
```
Comment (DianQK) [inline tests/assembly/issues/issue-126585.rs:23]: https://rust.godbolt.org/z/s9arM1vcW

File: tests/assembly/issues/issue-126585.rs [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+//@ compile-flags: -Copt-level=s -Cllvm-args=-x86-asm-syntax=intel
+//@ min-llvm-version: 19
+
+// Test for #126585.
+// Ensure that we do not create subsequent labels and unnecessary `jmp` instructions.
+
+#![crate_type = "lib"]
+
+#[no_mangle]
+fn checked_div_round(a: u64, b: u64) -> Option<u64> {
```
Comment (nikic) [inline tests/assembly/issues/issue-126585.rs:12]: I think it would be better to check that the IR does not have the extra undef phi input.

Comment (DianQK) [reply tests/assembly/issues/issue-126585.rs:12]: At least for this case, I think so.

File: tests/codegen/issues/issue-126585.rs [hunk 5]
```diff
@@ -0,0 +1,23 @@
+//@ compile-flags: -Copt-level=s
+//@ min-llvm-version: 19
+
+// Test for #126585.
+// Ensure that this IR doesn't has extra undef phi input, which also guarantees that this asm
+// doesn't has subsequent labels and unnecessary `jmp` instructions.
```
Comment (yotamofek) [inline tests/codegen/issues/issue-126585.rs:6]: Drive-by spellchecking 😄 

```suggestion
// Ensure that this IR doesn't have extra undef phi input, which also guarantees that this asm
// doesn't have subsequent labels and unnecessary `jmp` instructions.
```

Comment (DianQK) [reply tests/codegen/issues/issue-126585.rs]: Oops, thanks!

**Comment by DianQK** [comment]
I have already locally verified that test-various and arm-android have passed.

@bors try

**Comment by matthiaskrgr** [comment]
@bors r- 
https://github.com/rust-lang/rust/pull/128785#issuecomment-2273975638

**Comment by DianQK** [comment]
It appears to be caused by the ABI of i686, which should also be fixable. However, for this test, I'm not sure which is more appropriate: `ignore-x86`, `only-64bit`, or `only-x86_64`.

**Comment by nikic** [comment]
I'd go back to the original version without no-jump-table and only-x86_64.

**Comment by DianQK** [comment]
Sounds good, thanks, done at https://github.com/rust-lang/rust/compare/d432330181ba1d2cecc72238abde6d7abc3c1272..b5c453d7a28ab156dd6e2ed81674500cb1dbe7ec.
